src/main/java/com/advent/of/code/day6/LanternfishPredictor.java

```diff
@@ -0,0 +1,31 @@
+package com.advent.of.code.day6;
+
+import java.util.HashMap;
+import java.util.Map;
+
+class LanternfishPredictor {
+    long predictPopulationAfter(int atDay, Map<Integer, Long> individuals) {
+        final Map<Integer, Long> skipAtDay = new HashMap<>();
+        for (int day = 0; day < atDay; day++) {
+            final int dayInCycle = (day % 7) + 1;
+            final Map<Integer, Long> updated = new HashMap<>();
+            for (int key : individuals.keySet()) {
+                if (key - dayInCycle == -1) {
+                    final int newKey = (key + 2) % 7;
+                    final long currentValue = individuals.getOrDefault(newKey, 0L);
+                    final long newValue = currentValue + individuals.get(key) - skipAtDay.getOrDefault(day, 0L);
+                    updated.put(key, individuals.get(key));
+                    updated.put(newKey, newValue);
+                    skipAtDay.put(day + 2, individuals.get(key) - skipAtDay.getOrDefault(day, 0L));
+                } else {
+                    if (!updated.containsKey(key)) {
+                        updated.put(key, individuals.get(key));
+                    }
+                }
+            }
+            individuals = updated;
+        }
+        return individuals.values().stream().mapToLong(Long::longValue).sum();
```

Comment:
Ciekawe rozwiązanie: niby podobna konstrukcja (u mnie tablica long, a u Ciebie mapa <Integer, Long>, ale bardziej złożona logika po Twojej stronie. Bardzo mi się podoba jednowierszowe wyliczenie wyniku, zamiast stosowania jakichś pętli :)

Reply:
no z tym bym był ostrożny - tak jak ostatnio mówiliśmy wcale nie zawsze ten kod jest bardziej czytelny/zrozumiały :D dodatkowo jeszcze Java ma ten problem, że trzeba często robić kupę jakichś operacji dziwnych w stylu właśnie mapowania do LongStream ewentualnie możnaby tu zrobić `individuals.values().stream().collect(Collectors.summarizingLong(Long::longValue));` co dalej jest jakąś dziwną konstrukcją "na około" - przecież mam listę liczb, czemu nie mogę ich zsumować tylko muszę robić jakieś dziwne transformacje, rzutowania itp :D 

Tutaj Kotlin znowu ma trochę przewagę bo udostępnia masę funkcji pomocnicznych (jak właśnie `sum()` dla kolekcji liczb - https://pl.kotl.in/Vfsb-jWoj) które oczywiście są tym co ja wyżej musiałem zrobić ale zapakowane do biblioteki i user już nie musi tego boiler plate code pisać :)